devito/ir/clusters/analysis.py

```diff
@@ -75,32 +75,48 @@ def callback(self, clusters, prefix):
         # Apply the actual callback
         retval = self._callback(clusters, d, prefix)
 
+        # Normalize retval
+        retval = set(as_tuple(retval))
+
         # Update `self.state`
-        if retval is not None:
+        if retval:
             for c in clusters:
                 properties = self.state.properties.setdefault(c, {})
-                properties.setdefault(d, set()).add(retval)
+                properties.setdefault(d, set()).update(retval)
 
         return clusters
 
 
 class Parallelism(Detector):
 
     """
-    Detect SEQUENTIAL, PARALLEL, and PARALLEL_IF_ATOMIC Dimensions.
+    Detect SEQUENTIAL, PARALLEL, PARALLEL_INDEP and PARALLEL_IF_ATOMIC Dimensions.
+
+    Consider an IterationSpace over `n` Dimensions. Let `(d_1, ..., d_n)` be the
+    distance vector of a dependence. Let `i` be the `i-th` Dimension of the
+    IterationSpace. Then:
+
+        * `i` is PARALLEL_INDEP if all dependences have distance vectors:
+
+            (d_1, ..., d_i) = 0
+
+        * `i` is PARALLEL if all dependences have distance vectors:
```

Comment:
There's an OR here. This one is weaker than the other. Why do you say "same definition" ?

Reply:
In the equation, but the text is exactly the same for all three case that is super confusing because it looks like these property are true for the same cases.

Reply:
I'm not sure I'm following sorry. Could you treat me like a baby and tell me how you think I should rephrase it?